src/common/health.rs

```diff
@@ -0,0 +1,337 @@
+                "Not enough cluster nodes responded to GetConsensusCommit request: \
+                 required {required_commit_indices_count},
+                 responded {} out of {}",
+                commit_indices.len(),
+                peer_address_by_id.len(),
+            );
+
+            return None;
+        }
+
+        let cluster_commit_index = commit_indices
+            .into_iter()
+            .map(|resp| resp.into_inner().commit)
+            .max()
+            .unwrap_or(0);
+
+        Some(cluster_commit_index as _)
+    }
+
+    fn commit_index(&self) -> u64 {
+        // TODO: Blocking call in async context!?
+        self.consensus_state
+            .persistent
+            .read()
+            .state
+            .hard_state
+            .commit
+    }
+
+    async fn unhealthy_shards(&self) -> HashSet<Shard> {
+        let this_peer_id = self.toc.this_peer_id;
+        let collections = self.toc.all_collections().await;
+
+        let mut unhealthy_shards = HashSet::new();
+
+        for collection in &collections {
+            let state = match self.toc.get_collection(collection).await {
+                Ok(collection) => collection.state().await,
+                Err(_) => continue,
+            };
+
+            for (&shard, info) in state.shards.iter() {
+                let Some(state) = info.replicas.get(&this_peer_id) else {
+                    continue;
+                };
+
+                if is_shard_ready(state) {
+                    continue;
+                }
+
+                unhealthy_shards.insert(Shard::new(collection, shard));
+            }
+        }
+
+        unhealthy_shards
+    }
+
+    fn set_ready(&self) {
+        self.is_ready.store(true, atomic::Ordering::Relaxed);
+        self.is_ready_signal.notify_one();
+    }
+}
+
+fn downcast_str(any: &dyn any::Any) -> Option<&str> {
+    if let Some(str) = any.downcast_ref::<&'static str>() {
+        return Some(str);
+    }
+
+    if let Some(str) = any.downcast_ref::<String>() {
+        return Some(str);
+    }
+
+    None
+}
+
+fn get_consensus_commit<'a>(
+    transport_channel_pool: &'a TransportChannelPool,
+    uri: &'a tonic::transport::Uri,
+) -> impl Future<Output = GetConsensusCommitResult> + 'a {
+    transport_channel_pool.with_channel_timeout(
+        uri,
+        |channel| async {
+            let mut client = QdrantInternalClient::new(channel);
+            let mut request = tonic::Request::new(GetConsensusCommitRequest {});
+            request.set_timeout(defaults::CONSENSUS_META_OP_WAIT);
+            client.get_consensus_commit(request).await
+        },
+        Some(defaults::CONSENSUS_META_OP_WAIT),
+        2,
+    )
+}
+
+type GetConsensusCommitResult = Result<
+    tonic::Response<GetConsensusCommitResponse>,
+    transport_channel_pool::RequestError<tonic::Status>,
+>;
+
+fn is_shard_ready(state: &ReplicaState) -> bool {
+    matches!(state, ReplicaState::Active | ReplicaState::Listener)
```

Comment:
Should we prefer an exhaustive match here, or move the function into `ReplicaState`?

Reply:
I recently added a function similar to this, maybe we can add another one for this check below it: https://github.com/qdrant/qdrant/blob/ef9a11c200f773765ea2a5424226a761e5fc6db2/lib/collection/src/shards/replica_set/mod.rs#L882-L893

Reply:
Fixed.